replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/modern/RepSyncImpl.java

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ditto.replication.api.impl.modern;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.codice.ditto.replication.api.modern.RepSync;
+
+public class RepSyncImpl extends AbstractPersistable implements RepSync {
+
+  public static final String PERSISTENCE_TYPE = "replication_repsync";
+
+  private static final String NAME = "name";
+
+  private static final String SOURCE = "source";
+
+  private static final String DESTINATION = "destination";
+
+  private static final String FILTER = "filter";
+
+  public RepSyncImpl(String name, String source, String destination, String filter) {
+    this(null, name, source, destination, filter);
+  }
+
+  public RepSyncImpl(String id, String name, String source, String destination, String filter) {
+    super(id, PERSISTENCE_TYPE);
+    safePut(NAME, name);
+    safePut(SOURCE, source);
+    safePut(DESTINATION, destination);
+    safePut(FILTER, filter);
+  }
+
+  public RepSyncImpl(Map<String, Object> properties) {
+    super(PERSISTENCE_TYPE, properties);
+  }
+
+  @Override
+  public String getName() {
+    return (String) properties.get(NAME);
+  }
+
+  public void setName(String name) {
+    properties.put(NAME, name);
+  }
+
+  @Override
+  public String getSource() {
+    return (String) properties.get(SOURCE);
+  }
+
+  public void setSource(String sourceId) {
+    properties.put(SOURCE, sourceId);
+  }
+
+  @Override
+  public String getDestination() {
+    return (String) properties.get(DESTINATION);
+  }
+
+  public void setDestination(String destinationId) {
+    properties.put(DESTINATION, destinationId);
+  }
+
+  @Override
+  public String getFilter() {
+    return (String) properties.get(FILTER);
+  }
+
+  public void setFilter(String filter) {
+    properties.put(FILTER, filter);
+  }
+
+  @Override
+  public void setProperties(Map<String, Object> properties) {
+    if (!hasRequiredFields(properties)) {
+      throw new IllegalArgumentException("Null values are not allowed.");

```

Comment:
✏️ If there is an expectation that an implementation throws an `IllegalArgumetException` for a certain reason than this should be documented in the parent class along with the contract for this method

Reply:
❓ What about if the values for specific keys are of the wrong type? Deserialization, or here validation, of the object should ensure the object is completely valid before allowing the object to be used.